### DISCUSSION TITLE
Issues with DAB radio metadata

### DISCUSSION THREAD

**Discussion by vale-max** [rern/rAudio#1022] [Bug Reports]
Hi @rern 
tell me when you get enough of my posts about DAB radio :grin: 

I noticed that the station name was not picked up when playing the DAB radios. The problem is that when status.sh builds /srv/http/data/shm/radio, it looks for the radio file inside $dirwebradio which is /srv/http/data/webradio , but the DAB radio files are inside a different path so the file can't be found and the $station field is not valorized.

My simple fix has been to change line 297 in status.sh to:
```
                 [[ ! -e $radiofile  ]] && radiofile=$( find /srv/http/data -name "$urlname" )
```

Another problem is with the DAB label and slide name.
The text inside the label can be really anything and it sometimes confuses the parsing and rename of the slide file.

An example, at some moment the image changed to the default rAudio logo, this was the label in that moment:
``Radio Italia SMI - viale Europa 49 - 20093 Cologno Monzese (MI)``
This the coverart file name in status file:
``coverart="/data/shm/webradio/RadioItaliaSMIvialeEuropa49-20093ColognoMonzese(MI).jpg"``
but this file didn't exist:
```
ls -lrt /srv/http/data/shm/webradio/
total 96
-rw-r--r-- 1 root root 6955 Sep 12 21:06  VirginRadioStyleRock.jpg
-rw-r--r-- 1 root root 7271 Sep 12 21:11  REDHOTCHILIPEPPERSTIPPAMYTONGUE.jpg
-rw-r--r-- 1 root root 6964 Sep 12 21:15  PINKFLOYDTHEGREATGIGINTHESKY.jpg
-rw-r--r-- 1 root root 6955 Sep 12 21:19  ACDCBIGJACK.jpg
-rw-r--r-- 1 root root 8513 Sep 12 21:32  RadioItaliaemail:diretta@radioitalia.it.jpg
-rw-r--r-- 1 root root 8513 Sep 12 21:37  RadioItaliatel.02-25.44.41.jpg
-rw-r--r-- 1 root root 8513 Sep 12 21:38 'RADIOITALIALIVEinondaogniVenerd'$'\303\254''alle21.00eincontemporaneasuRadioItaliaTv(canale70delDTT)einstreaming.jpg'
-rw-r--r-- 1 root root 8513 Sep 12 21:40  RadioItaliasmseWhatsApp:339.99.500.95.jpg
-rw-r--r-- 1 root root 8513 Sep 12 21:41  DABslide.jpg
-rw-r--r-- 1 root root   63 Sep 12 21:41  DABlabel.txt
```
I did not try to debug what was the failure in the file name creation, IMHO it is too much effort, I just changed the file name creation using a timestamp.
Also trying to separate Album and title name works only on some occasions, most of the times the text is totally random, some radio sends news headlines as label.

For me this is what works better:
```
--- status-radio-original.sh    2022-09-12 23:05:08.634324164 +0200
+++ status-radio.sh     2022-09-13 17:32:21.017066278 +0200
@@ -39,13 +39,8 @@
 esac

 dabData() {
-       artist_title=$( sed -E 's/ - |: /\n/' $dirshm/webradio/DABlabel.txt )
-       if [[ $artist_title == $( head -2 $dirshm/status ) ]]; then
-               sleep 10
-               metadataGet
-       else
-               readarray -t metadata <<< "$artist_title"
-       fi
+       DABlab="$(cat /srv/http/data/shm/webradio/DABlabel.txt)"
+       metadata=("$station" "${DABlab//\"/}" "DAB radio" "dab" "10")
 }
 radiofranceData() {
        readarray -t metadata <<< $( curl -sGk -m 5 \
@@ -105,7 +100,12 @@
                if [[ $coverurl != dab ]]; then
                        curl -s $coverurl -o $coverfile
                else
-                       mv $dirshm/webradio/{DABslide,$name}.jpg
+                       slidename=DABslide$( date +%s%3N ).jpg
+                       coverart=/data/shm/webradio/$slidename
+                       cp /srv/http/data/shm/webradio/DABslide.jpg /srv/http/data/shm/webradio/$slidename
+                       touch /srv/http/data/shm/webradio/DABlabel.txt #avoid label deletion by coverfileslimit
+                       touch /srv/http/data/shm/webradio/DABslide.jpg #avoid label deletion by coverfileslimit
+
                fi
        else
                coverart=$( ls $dirshm/webradio/$name* 2> /dev/null | sed 's|/srv/http||' )

```

**Comment by rern** [comment]
Please try today's update.

**Comment by vale-max** [comment]
@rern there's something wrong, sometimes the coverart works, sometimes changes to the R logo.
This errors shows in the journal
```
Sep 17 19:44:38 rAudio nginx[662]: 2022/09/17 19:44:38 [error] 662#662: *35 open() "/srv/http/data/shm/webradio/DABslide.jpg" failed (2: No such file or directory), client: 127.0.0.1, server: localhost, request: "GET /data/shm/webradio/DABslide.1663436677.jpg HTTP/1.1", host: "localhost", referrer: "http://localhost/"
```
I'm not sure why but nginx or something else seems to strip the date string from the file name.
Using
```
                       coverart=/data/shm/webradio/DABslide${date}.jpg
#                       rm -f $dirshm/webradio/DABslide.*.jpg
                        mv $dirshm/webradio/DABslide{,$date}.jpg
```
almost solves the problem but still sometimes:
```
Sep 17 20:12:47 rAudio nginx[662]: 2022/09/17 20:12:47 [error] 662#662: *35 open() "/srv/http/data/shm/webradio/DABslide1663438367.jpg" failed (2: No such file or directory), client: 127.0.0.1, server: localhost, request: "GET /data/shm/webradio/DABslide1663438367.jpg HTTP/1.1", host: "localhost", referrer: "http://localhost/"
```
I think this depends because the code is executed every time metadataget detects that the label has changed, in this case sets the new coverart name then moves the new slide to the new name, but the slide is sent in a way totally independent from the label so it can happen that there is no new file to be moved thus the new file will not exist but coverart name already changed.

It is better not to move DABslide.jpg but copying it, so in case it did not change (some radios send it very often, some update the image every now and then, some other do not send any at all) the last one is always copied to the expected name one, like:
```
               if [[ $icon == dabradio ]]; then
                        date=$( date +%s )
                        coverart=/data/shm/webradio/DABslide${date}.jpg
                        rm -f $dirshm/webradio/DABslide??????????.jpg
                        cp $dirshm/webradio/DABslide{,$date}.jpg
                else
```

There is still the possibility that the browser tries to retrieve the image in that short moment when the old file has been deleted and the name has not been updated but that's bad luck...

Also, the fact that the label did not change does not necessarily mean that the slide is also unchanged, Virgin Radio for example changes slides within the same song, so running the cover art update code only if the label text changed is conceptually wrong (also now does not work because the string taken from the status file has quotes and is compared with a string without quotes)

So either you compare like ``if [[ $( grep ^Title $dirshm/status | cut -d= -f2- ) == "\"$dablabel\"" ]]; then`` or better remove all that part and do:
```
        else
                icon=dabradio
                dablabel=$( cat $dirshm/webradio/DABlabel.txt )
                metadata=( "$station" "$dablabel" 'DAB Radio' '' 10 )
        fi
 ```
Last detail, as the slide is copied but not moved, dab-start.sh should delete it when it exits
```
killsubs() {
        kill $DABPID
        kill $FFMPID
        rm $MYPIPE
        echo NO INFO > /srv/http/data/shm/webradio/DABlabel.txt
        rm /srv/http/data/shm/webradio/DABslide.jpg
}
```

**Reply by rern** [reply]
@vale-max,

Please try the UPDATE branch:
- `dab.service` > `status-dab.sh` instead of `radio.service` > `status-radio.sh`
- Refresh only when `DABlabel.txt` changed instead of polling every 15s

**Reply by vale-max** [reply]
Hi @rern 
several problems, the coverart name was missing the DABslide part, the label was displayed in the browser without spaces, in the sampling the position was displayed as $pos, the coverart name was updated but the file to be moved was not existing yet so R logo displayed and probably also issues with the rename like in:
![Screenshot from 2022-09-18 22-01-48](https://user-images.githubusercontent.com/54400080/190929357-44a7f6cd-8d38-4cdf-b798-322a48efce6b.png)

The refresh rate with inotify I think is excessive, the new status is sent almost always every 2 seconds and sometimes more than one refresh was sent within the same second.
Frankly I think all gets too complicated and the previous code with the dab radio status update as part of status-radio.sh, with my last suggestions, was working fine, the slide was always displayed and the text was fine also in case of very long labels.
Refreshing very often also makes difficult to edit the playlist as the view is bounced to the current song played every time a status update is received.

If you want to keep the separate status script for DAB however the adjustments below make it work fine:
```
--- status-dab.sh_orig	2022-09-18 21:50:41.918765466 +0200
+++ status-dab.sh	2022-09-19 00:00:17.492438026 +0200
@@ -10,10 +10,14 @@
 filelabel=$dirshm/webradio/DABlabel.txt
 touch $filelabel
 
-while read line; do
-	title=$( cat $filelabel | tr -d ' \"`?/#&'"'" )
-	coverart=/data/shm/webradio/$title.jpg
-	mv $dirshm/webradio/DABslide{,$title}.jpg &> /dev/null
+while true; do
+	title=$( cat $filelabel )
+	filedate=$( date +%s )
+	coverart=/data/shm/webradio/DABslide${filedate}.jpg	
+	cp $dirshm/webradio/DABslide{,$filedate}.jpg &> /dev/null
+	touch $dirshm/webradio/DABslide.jpg
+	touch $filelabel
+	sampling="$pos &bull; 48 kHz 160 kbit/s"
 	elapsed=$( printf '%.0f' $( { echo status; sleep 0.05; } \
 				| telnet 127.0.0.1 6600 2> /dev/null \
 				| grep ^elapsed \
@@ -25,7 +29,7 @@
 , "elapsed"  : '$elapsed'
 , "file"     : "'$file'"
 , "icon"     : "dabradio"
-, "sampling" : "$pos &bull; 48 kHz 160 kbit/s"
+, "sampling" : "'$sampling'"
 , "state"    : "play"
 , "song"     : '$song'
 , "station"  : ""
@@ -33,4 +37,6 @@
 , "Title"    : "'$title'"
 }'
 	$dirbash/status-push.sh statusradio "$data" &
-done < <( inotifywait -mq -e close_write $filelabel )
+	sleep 10
+done
+
```
Add to this the deletion of the slide file when dab-start.sh is stopped:
```
killsubs() {
        kill $DABPID
        kill $FFMPID
        rm $MYPIPE
        echo NO INFO > /srv/http/data/shm/webradio/DABlabel.txt
        rm /srv/http/data/shm/webradio/DABslide.jpg
}
```

**Reply by rern** [reply]
Please try the revised UPDATE branch again.

**Reply by vale-max** [reply]
still some glitch, at a point the image changed to the R logo and remained as R for a long time, this was the directory content and the status file during this phase:
```
rAudio:~  #  cat /srv/http/data/shm/status
Album="DAB Radio"
Artist="#RTL 102.5"
coverart=""
elapsed=555
file="rtsp://rAudio:8554/12a_5218"
icon="dabradio"
sampling="1/36 • 48 kHz 160 kbit/s • DAB"
state="play"
song=0
station=""
Time=false
Title="_ spettacolo per le scuole, una scala esterna ha ceduto e 13 tra studenti e professori sono rimasti feriti. Ricoverati"
timestamp=1663884582075
webradio=true
player=mpd
rAudio:~  #  ls -l /srv/http/data/shm/webradio
total 72
-rw-r--r-- 1 root root  118 Sep 23 00:10  DABlabel.txt
-rw-r--r-- 1 root root 7430 Sep 23 00:09  DABslide.jpg
-rw-r--r-- 1 root root  118 Sep 23 00:09  DABtitle
-rw-r--r-- 1 root root 7430 Sep 23 00:09  FrancescoGabbani-Amen.jpg
-rw-r--r-- 1 root root 9831 Sep 23 00:08  INottambuliconAngeloDiBenedettoeAndreaDeSabato.jpg
-rw-r--r-- 1 root root 7430 Sep 23 00:09 'News:ititolidioggivenerd'$'\303\254''23settembre2022.jpg'
-rw-r--r-- 1 root root 7430 Sep 23 00:09  PauraalGlobeTheatre,ilteatroshakespearianonatodaunideadiGigiProiettiaVillaBorgheseaRoma.Allafinediuno_.jpg
-rw-r--r-- 1 root root 9831 Sep 23 00:08  Rema-CalmDown.jpg
-rw-r--r-- 1 root root 7430 Sep 23 00:09  _spettacoloperlescuole,unascalaesternahacedutoe13trastudentieprofessorisonorimastiferiti.Ricoverati.jpg
```

Then at some point the image came back:
```
rAudio:~  #  cat /srv/http/data/shm/status
Album="DAB Radio"
Artist="#RTL 102.5"
elapsed=696
file="rtsp://rAudio:8554/12a_5218"
icon="dabradio"
sampling="1/36 • 48 kHz 160 kbit/s • DAB"
state="play"
song=0
station=""
Time=false
Title="I Nottambuli con Angelo Di Benedetto e Andrea De Sabato"
timestamp=1663884723141
webradio=true
player=mpd
coverart=/data/shm/webradio/INottambuliconAngeloDiBenedettoeAndreaDeSabato.1663884733.jpg
rAudio:~  #  ls -l /srv/http/data/shm/webradio
total 72
-rw-r--r-- 1 root root  7430 Sep 23 00:11  AlviadomaniireferendumperlannessioneallaRussianelleregioniucraineoccupate.Sonoillegali,diceilsegretario_.jpg
-rw-r--r-- 1 root root    55 Sep 23 00:12  DABlabel.txt
-rw-r--r-- 1 root root 10220 Sep 23 00:12  DABslide.jpg
-rw-r--r-- 1 root root    55 Sep 23 00:12  DABtitle
-rw-r--r-- 1 root root 10220 Sep 23 00:12  INottambuliconAngeloDiBenedettoeAndreaDeSabato.jpg
-rw-r--r-- 1 root root  7430 Sep 23 00:11  TremamezzaItalia.Scossediterremoto,conmagnitudofra3,6e4,1,sisonoregistratedaGenovaaCatania.Avvertitein_.jpg
-rw-r--r-- 1 root root  7430 Sep 23 00:11  _Liguria,Emilia-Romagna,Marche,AbruzzoeSicilia.Paura,manonsisegnalanovittimeodannidirilievo.Secondol.jpg
-rw-r--r-- 1 root root  7430 Sep 23 00:10  _abbiamodeglistrumenti,comenelcasodiPoloniaeUngheria.LohadettolapresidentedellaCommissioneeu.jpg
-rw-r--r-- 1 root root  7430 Sep 23 00:11 '_generaledellOnuGuterres.IlvicepresidenteMedvedevribadiscechedopoireferendumMosca'$'\303\250''prontaausareanchear.jpg'
```

Sorry, I don't have time now to debug what's wrong
